### PR TITLE
Make podman generate systemd --new flag parsing more robust

### LINKED ISSUE #8847: Bug: podman generate systemd defines multiple "-d"
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

Podman generates multiple `-d` paramater in the created output of `podman generate systemd <container>`, when -dt is already defined.

**Steps to reproduce the issue:**

1. `podman run -dt --name web01 httpd:2`
2. `podman generate systemd --new web01`

**Describe the results you received:**

Check the `ExecStart=` Part in the below code. it has an additional `-d`

```
# container-14978c6f5f2311d850accd2803c305473cf0d02170ea65f7132bbc41a89bd7d9.service
# autogenerated by Podman 2.2.1
# Mon Dec 28 16:49:38 CET 2020

[Unit]
Description=Podman container-14978c6f5f2311d850accd2803c305473cf0d02170ea65f7132bbc41a89bd7d9.service
Documentation=man:podman-generate-systemd(1)
Wants=network.target
After=network-online.target

[Service]
Environment=PODMAN_SYSTEMD_UNIT=%n
Restart=on-failure
ExecStartPre=/bin/rm -f %t/container-14978c6f5f2311d850accd2803c305473cf0d02170ea65f7132bbc41a89bd7d9.pid %t/container-14978c6f5f2311d850accd2803c305473cf0d02170ea65f7132bbc41a89bd7d9.ctr-id
ExecStart=/usr/bin/podman run --conmon-pidfile %t/container-14978c6f5f2311d850accd2803c305473cf0d02170ea65f7132bbc41a89bd7d9.pid --cidfile %t/container-14978c6f5f2311d850accd2803c305473cf0d02170ea65f7132bbc41a89bd7d9.ctr-id --cgroups=no-conmon -d --replace -dt --name web01 httpd:2
ExecStop=/usr/bin/podman stop --ignore --cidfile %t/container-14978c6f5f2311d850accd2803c305473cf0d02170ea65f7132bbc41a89bd7d9.ctr-id -t 10
ExecStopPost=/usr/bin/podman rm --ignore -f --cidfile %t/container-14978c6f5f2311d850accd2803c305473cf0d02170ea65f7132bbc41a89bd7d9.ctr-id
PIDFile=%t/container-14978c6f5f2311d850accd2803c305473cf0d02170ea65f7132bbc41a89bd7d9.pid
KillMode=none
Type=forking

[Install]
WantedBy=multi-user.target default.target
```

**Describe the results you expected:**

Shorthandles from `podman run` are expanded to long handles (-e = --env, -d = --detach, etc.) and duplicates are avoided or removed.

**Additional information you deem important (e.g. issue happens only occasionally):**

Does not occure if only `-d` is given during run.

**Output of `podman version`:**

```
Version:      2.2.1
API Version:  2.1.0
Go Version:   go1.15.5
Built:        Tue Dec  8 15:37:50 2020
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.18.0
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.0.21-3.fc33.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.21, commit: 0f53fb68333bdead5fe4dc5175703e22cf9882ab'
  cpus: 8
  distribution:
    distribution: fedora
    version: "33"
  eventLogger: journald
  hostname: nb01.while-true-do.io
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 5.9.15-200.fc33.x86_64
  linkmode: dynamic
  memFree: 265486336
  memTotal: 16493572096
  ociRuntime:
    name: crun
    package: crun-0.16-3.fc33.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 0.16
      commit: eb0145e5ad4d8207e84a327248af76663d4e50dd
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    path: /run/podman/podman.sock
  rootless: false
  slirp4netns:
    executable: ""
    package: ""
    version: ""
  swapFree: 3980914688
  swapTotal: 4294963200
  uptime: 27h 44m 19.34s (Approximately 1.12 days)
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - registry.centos.org
  - docker.io
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 7
    paused: 0
    running: 7
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.mountopt: nodev,metacopy=on
  graphRoot: /var/lib/containers/storage
  graphStatus:
    Backing Filesystem: btrfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "true"
  imageStore:
    number: 8
  runRoot: /var/run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 2.1.0
  Built: 1607438270
  BuiltTime: Tue Dec  8 15:37:50 2020
  GitCommit: ""
  GoVersion: go1.15.5
  OsArch: linux/amd64
  Version: 2.2.1
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-2.2.1-1.fc33.x86_64
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**

### PR DESCRIPTION
**Pull request by Luap99**
First, use the pflag library to parse the flags. With this we can
handle all corner cases such as -td or --detach=false.

Second, preserve the root args with --new. They are used for all podman
commands in the unit file. (e.g. podman --root /tmp run alpine)

Fixes #8847

### CODE REVIEW

**Comment by rhatdan** [comment]
Is there a corner case where a user could specify something like

podman run -d --detach=false alpine sleep 100

And we end up with using the first -d?

**Comment by Luap99** [comment]
`-d -detach=false alpine sleep 100` would turn into `-d -d alpine sleep 100` which is valid

**Comment by rhatdan** [comment]
It is valid but not what the user requested.

podman run -d --detach=false alpine sleep 100
Is treated as
podman run --detach=false alpine sleep 100

With your change it will be treated as 
podman run --detach=true alpine sleep 100

BTW I have no problem if we just make running multiple --detach commands fail.

**Comment by Luap99** [comment]
The whole point of the --new flag is that we must add detach regardless of what the user requested. If there is no detach systemctl start will fail on that unit.

**Comment by rhatdan** [comment]
Should it fail then if the user requested it not to detach?

**Comment by alitvak69** [comment]
Sorry for butting in but I don't think systemd will fail with no detach.  Here is my example that runs well.

`
[Unit]
Description=DNS Bind container
After=multi-user.target

[Service]
EnvironmentFile=-/etc/environment
Type=simple
#LimitNOFILE=32768
WorkingDirectory=/dns/slave-internal
ExecStartPre=-/usr/bin/podman pull docker.xcastlabs.net/services/bind-host:latest
ExecStart=/usr/bin/podman run --rm --name=dns-slave-internal --no-hosts=true -t --cgroups=no-conmon \
--log-driver=journald --log-opt=tag={{.Name}} \
 --env=CONTAINER_HOST=container2-chi.xcastlabs.net \
    --env=DC=chi \
    --env=DNSZONE_TYPE=slave \
    --env=DNSZONE_VIEW=internal \
    --env=SERVICE_NAME=dns-slave-internal \
    --env=SERVICE_PORT=53 \
    --env=BIND_DATA_DIR=/dns \
    --cap-add=CAP_SYS_PTRACE,CAP_NET_ADMIN,CAP_NET_RAW,CAP_NET_BIND_SERVICE,CAP_SETUID,CAP_SETGID \
    --volume=/dns/slave-internal:/dns:rw \
    --net=private-dns-internal-slaves \
     \
     docker.xcastlabs.net/services/bind-host:latest
ExecStop=-/usr/bin/podman stop dns-slave-internal
SyslogIdentifier=dns-slave-internal
Restart=on-failure
KillMode=none
RestartSec=10s
TimeoutStartSec=120
TimeoutStopSec=15

[Install]
WantedBy=multi-user.target
dns-slave-internal.service (END)
`

**Comment by Luap99** [comment]
Yes it works when the unit type is simple but we set the type to forking.

File: pkg/systemd/generate/containers_test.go [hunk 1]
```diff
@@ -304,7 +405,7 @@ WantedBy=multi-user.target default.target
 				PIDFile:           "/var/run/containers/storage/overlay-containers/639c53578af4d84b8800b4635fa4e680ee80fd67e0e6a2d4eea48d1e3230f401/userdata/conmon.pid",
 				StopTimeout:       42,
 				PodmanVersion:     "CI",
-				CreateCommand:     []string{"I'll get stripped", "container", "run", "--name", "jadda-jadda", "--hostname", "hello-world", "awesome-image:latest", "command", "arg1", "...", "argN", "foo=arg \"with \" space"},
+				CreateCommand:     []string{"I'll get stripped", "run", "--name", "jadda-jadda", "--hostname", "hello-world", "awesome-image:latest", "command", "arg1", "...", "argN", "foo=arg \"with \" space"},
```
Comment (vrothberg) [inline pkg/systemd/generate/containers_test.go:408]: Please leave at least one instance with "container".

**Review by vrothberg — COMMENTED** [review]
Thanks for tackling it, really nice work!

Please make sure to split future commits into separate pieces.  This commit applies too many changes at once which makes it hard to review and may hinder backporting.  But we're close to a new release, so we're good.

File: pkg/systemd/generate/containers_test.go (hunk 1 above)
Comment (Luap99) [reply pkg/systemd/generate/containers_test.go]: ok, done

File: test/e2e/generate_systemd_test.go [hunk 2]
```diff
@@ -59,8 +59,7 @@ var _ = Describe("Podman generate systemd", func() {
 		session = podmanTest.Podman([]string{"generate", "systemd", "--restart-policy", "bogus", "foobar"})
 		session.WaitWithDefaultTimeout()
 		Expect(session).To(ExitWithError())
-		found, _ := session.ErrorGrepString("bogus is not a valid restart policy")
-		Expect(found).Should(BeTrue())
+		Expect(session.ErrorToString()).To(ContainSubstring("bogus is not a valid restart policy"))
```
[the remark this answers is not in the document]
Comment (Luap99) [reply test/e2e/generate_systemd_test.go:62]: You are right, my bad. I was debugging the test failures and the  `Expect(found).Should(BeTrue())` syntax doesn't output useful information.

[the remark this answers is not in the document]
Comment (vrothberg) [reply test/e2e/generate_systemd_test.go:62]: No worries at all.  I love the improvements :+1:

**Comment by rhatdan** [comment]
LGTM

**Pull request merged**